src/randomenv.cpp

```diff
@@ -0,0 +1,389 @@
+}
+
+void AddFile(CSHA512& hasher, const char *path)
+{
+    struct stat sb;
+    int f = open(path, O_RDONLY);
+    if (f != -1) {
+        unsigned char fbuf[1024];
+        int n;
+        hasher.Write((const unsigned char*)&f, sizeof(f));
+        if (fstat(f, &sb) == 0) {
+            AddStat(hasher, sb);
+        }
+        do {
+            n = read(f, fbuf, sizeof(fbuf));
+            if (n > 0) hasher.Write(fbuf, n);
+            /* not bothering with EINTR handling. */
+        } while (n == sizeof(fbuf));
+        close(f);
+    }
+}
+
+void AddPath(CSHA512& hasher, const char *path)
+{
+    struct stat sb;
+    if (stat(path, &sb) == 0) {
+        hasher.Write((const unsigned char*)path, strlen(path) + 1);
+        AddStat(hasher, sb);
+    }
+}
+#endif
+
+void AddSysctl(CSHA512& hasher, const char* path)
+{
+#if HAVE_DECL_SYSCTLBYNAME
+    unsigned char buffer[4096];
+    size_t siz = 4096;
+    int ret = sysctlbyname(path, buffer, &siz, nullptr, 0);
+    if (ret == 0 || (ret == -1 && errno == ENOMEM)) {
+        hasher.Write((const unsigned char*)path, strlen(path) + 1);
+        hasher.Write(buffer, std::min<size_t>(siz, 4096));
+    }
+#endif
+}
+
+#ifdef HAVE_GETCPUID
+void AddCPUID(CSHA512& hasher)
+{
+    uint32_t ax, bx, cx, dx;
+    GetCPUID(0, 0, ax, bx, cx, dx);
+    hasher << ax << bx << cx << dx;
+    uint32_t max = ax;
+    for (uint32_t leaf = 1; leaf <= max; ++leaf) {
+        for (uint32_t subleaf = 0;; ++subleaf) {
+            GetCPUID(leaf, subleaf, ax, bx, cx, dx);
+            hasher << leaf << subleaf << ax << bx << cx << dx;
+            // Iterate over subleaves for leaf 4, 11, 13
+            if (leaf != 4 && leaf != 11 && leaf != 13) break;
+            if ((leaf == 4 || leaf == 13) && ax == 0) break;
+            if (leaf == 11 && (cx & 0xFF00) == 0) break;
+        }
+    }
+}
+#endif
+} // namespace
+
+void RandAddDynamicEnv(CSHA512& hasher)
+{
+    RandAddSeedPerfmon(hasher);
+
+#ifdef WIN32
+    FILETIME ftime;
+    GetSystemTimeAsFileTime(&ftime);
+    hasher << ftime;
+#else
+#  ifdef __MACH__
+    hasher << mach_absolute_time();
+#  else
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    hasher << ts.tv_sec << ts.tv_nsec;
+#  endif
+    struct timeval tv;
+    gettimeofday(&tv, nullptr);
+    hasher << tv.tv_sec << tv.tv_usec;
+
+    // Current resource usage.
+    struct rusage usage;
+    if (getrusage(RUSAGE_SELF, &usage) == 0) {
+        hasher << usage.ru_utime.tv_sec << usage.ru_utime.tv_usec << usage.ru_stime.tv_sec << usage.ru_stime.tv_usec;
+        hasher << usage.ru_maxrss << usage.ru_minflt << usage.ru_majflt << usage.ru_inblock << usage.ru_oublock;
+        hasher << usage.ru_nvcsw << usage.ru_nivcsw;
+    }
+
+#ifdef __linux__
+    AddFile(hasher, "/proc/diskstats");
+    AddFile(hasher, "/proc/vmstat");
+    AddFile(hasher, "/proc/schedstat");
```

Comment:
I think this should be `/proc/self/schedstat`

Reply:
It seems both exist.